### ISSUE TITLE
clippy::type_repetition_in_bounds mistakes plain generic for its slice

### ISSUE THREAD

**Issue by Tamschi** [rust-lang/rust-clippy#8457]
### Summary

I came across this today, so maybe it's one of the reasons this is turned off frequently.

### Lint Name

clippy::type_repetition_in_bounds

### Reproducer

I tried this code:

```rust
#![warn(clippy::pedantic)]

use serde::de::DeserializeOwned;
use std::{
    borrow::Cow,
    fmt::{self, Debug, Formatter},
};

pub struct IntegerData<'a, T>(pub Cow<'a, [T]>)
where
    [T]: ToOwned,
    <[T] as ToOwned>::Owned: DeserializeOwned;

impl<'a, T> Debug for IntegerData<'a, T>
where
    T: Debug + PartialEq,
    [T]: ToOwned,
    <[T] as ToOwned>::Owned: DeserializeOwned,
{
    fn fmt(&self, _: &mut Formatter<'_>) -> fmt::Result {
        unimplemented!()
    }
}
```

I saw this happen:

```
warning: this type has already been used as a bound predicate
  --> src\lib.rs:17:5
   |
17 |     [T]: ToOwned,
   |     ^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> src\lib.rs:1:9
   |
1  | #![warn(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::type_repetition_in_bounds)]` implied by `#[warn(clippy::pedantic)]`
   = help: consider combining the bounds: `[T]: Debug + PartialEq + ToOwned`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_repetition_in_bounds
```

I expected to see this happen:

I believe this is a false positive, since I can't shorten the bounds here without limiting function applicability or making the body (which depends on `T: Debug + PartialEq`) not compile.

The suggestion also is invalid in that regard.

### Version

```text
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-pc-windows-msvc
release: 1.58.1
LLVM version: 13.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by giraffate** [comment]
Can you reproduce it on the playground? I couldn't do it: <https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=06ed3d8710f44ddc63f2542d3918e109>.

**Comment by Tamschi** [comment]
> Can you reproduce it on the playground? I couldn't do it: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=06ed3d8710f44ddc63f2542d3918e109.

I can't reproduce it there and locally it only happens on stable, not beta or nightly, and only when running specifically `cargo clippy` instead of `cargo check`. I assume it's been fixed already and the change just hasn't propagated into the stable release yet.
